morpho5/datastructures/complex.c

```diff
@@ -0,0 +1,581 @@
+/** @file complex.c
+ *  @author Danny Goldstein
+ *
+ *  @brief Complex number type
+ */
+
+#include <string.h>
+#include "object.h"
+#include "complex.h"
+#include "morpho.h"
+#include "builtin.h"
+#include "veneer.h"
+#include "common.h"
+
+/* **********************************************************************
+ * Complex objects
+ * ********************************************************************** */
+
+objecttype objectcomplextype;
+
+/** Function object definitions */
+size_t objectcomplex_sizefn(object *obj) {
+    return sizeof(objectcomplextype)+sizeof(double) * 2;
+}
+
+void objectcomplex_printfn(object *obj) {
+    printf("<Complex>");
+}
+
+objecttypedefn objectcomplexdefn = {
+    .printfn=objectcomplex_printfn,
+    .markfn=NULL,
+    .freefn=NULL,
+    .sizefn=objectcomplex_sizefn
+};
+
+/** Creates a complex object */
+objectcomplex *object_newcomplex(double real,double imag) {
+    objectcomplex *new = (objectcomplex *) object_new(sizeof(objectcomplex), OBJECT_COMPLEX);
+    
+    if (new) {
+        new->real=real;
+        new->imag=imag;
+    }
+    
+    return new;
+}
+
+/* **********************************************************************
+ * Other constructors
+ * ********************************************************************** */
+
+/*
+ * Create complex number from a float
+ */
+
+/** Creates a new array from a float */
+objectcomplex *object_complexfromfloat(double val) {
+    objectcomplex *ret=object_newcomplex(val,0.0);
+    return ret;
+}
+
+/*
+ * Clone complex
+ */
+
+/** Clone a complex */
+objectcomplex *object_clonecomplex(objectcomplex *in) {
+    objectcomplex *new = object_newcomplex(in->real,in->imag);
+    return new;
+}
+
+/* **********************************************************************
+ * Complex operations
+ * ********************************************************************* */
+
+/** @brief Sets the real part of a complex number.*/
+void complex_setreal(objectcomplex *c, double value) {
+    c->real = value;
+}
+/** @brief Sets the imaginary part of a complex number.*/
+void complex_setimag(objectcomplex *c, double value) {
```

Comment:
Perhaps `setreal` and `setimag` might be useful methods to have for the end user as well?

Reply:
Perhaps, but I think that can be done with assignment and .real()  and .imag()

for example
var A = Complex(1,1)
var A = Complex(2,A.imag())